File: 45_jump_game_II.js
const prompt = require("prompt-sync")();
function minJumpsToEnd(nums) {
  let n = nums.length;
  if (n <= 1) return 0; // Already at the last index

  let queue = [[0, 0]]; // [current index, jumps]
  let visited = new Set(); // To avoid revisiting indexes

  while (queue.length > 0) {
    let [index, jumps] = queue.shift();

    // Explore all possible jumps from the current index
    for (let step = 1; step <= nums[index]; step++) {
      let newIndex = index + step;

      if (newIndex >= n - 1) return jumps + 1; // Reached last index

      if (!visited.has(newIndex)) {
        visited.add(newIndex);
        queue.push([newIndex, jumps + 1]);
      }
    }
  }

  return -1; // If it's not possible to reach the last index
}

// Function to take user input and validate constraints
function getUserInput() {
  let input = prompt("Enter space-separated numbers:");
  let nums = input.split(" ").map(Number);

  // Validate constraints
  if (nums.length < 1 || nums.length > 1000) {
    console.log("Length of the array must be between 1 and 1000.");
    return;
  }
  if (nums.some((num) => num < 0 || num > 100000)) {
    console.log("Each number must be between 0 and 100000.");
    return;
  }

  let result = minJumpsToEnd(nums);
  console.log("Minimum jumps needed:", result);
}

// Call function to take input and compute result
getUserInput();

File: 45_jump_game_II.py
from collections import deque

def min_jumps_to_end(nums):
    n = len(nums)
    if n <= 1:
        return 0  # Already at the last index
    
    queue = deque([(0, 0)])  # (current index, jumps)
    visited = set()  # To avoid revisiting indexes

    while queue:
        index, jumps = queue.popleft()

        # Explore all possible jumps from the current index
        for step in range(1, nums[index] + 1):
            new_index = index + step

            if new_index >= n - 1:  # Reached the last index
                return jumps + 1
            
            if new_index not in visited:
                visited.add(new_index)
                queue.append((new_index, jumps + 1))

    return -1  # If it's not possible to reach the last index

# Taking user input with constraints
while True:
    try:
        nums = list(map(int, input("Enter space-separated numbers: ").split()))
        
        # Check constraints
        if not (1 <= len(nums) <= 1000):
            raise ValueError("Length of the array must be between 1 and 1000.")
        if any(num < 0 or num > 10**5 for num in nums):
            raise ValueError("Each number must be between 0 and 100000.")
        
        result = min_jumps_to_end(nums)
        print("Minimum jumps needed:", result)
        break
    except ValueError as e:
        print(f"Invalid input: {e}. Please try again.")
